packages/flutter_tools/lib/src/android/gradle.dart

```diff
@@ -211,445 +192,285 @@ void createSettingsAarGradle(Directory androidDirectory) {
+      command,
+      workingDirectory: project.android.hostAppGradleRoot.path,
+      allowReentrantFlutter: true,
+      environment: gradleEnv,
+      mapFunction: (String line) {
+        // This message was removed from 1P plugins,
+        // but older plugin versions still display this message.
+        if (androidXPluginWarningRegex.hasMatch(line)) {
+          // Don't pipe.
+          return null;
+        }
+        if (detectedGradleError == null) {
+          for (final GradleHandledError gradleError in gradleErrors) {
+            if (gradleError.test(line)) {
+              detectedGradleErrorLine = line;
+              detectedGradleError = gradleError;
+            }
+          }
+        }
+        // Pipe stdout/sterr from Gradle.
+        return line;
+      },
+    );
+  } finally {
+    status.stop();
   }
 
-  final File localProperties = project.android.localPropertiesFile;
-  bool changed = false;
-
-  SettingsFile settings;
-  if (localProperties.existsSync()) {
-    settings = SettingsFile.parseFromFile(localProperties);
-  } else {
-    settings = SettingsFile();
-    changed = true;
+  if (exitCode != 0 && detectedGradleError == null) {
+    BuildEvent('unknown-failure').send();
+    throwToolExit(
+      'Gradle task $assembleTask failed with exit code $exitCode',
+      exitCode: exitCode,
+    );
   }
 
-  void changeIfNecessary(String key, String value) {
-    if (settings.values[key] != value) {
-      if (value == null) {
-        settings.values.remove(key);
-      } else {
-        settings.values[key] = value;
-      }
-      changed = true;
+  if (exitCode != 0 && detectedGradleError != null) {
+    final GradleBuildStatus status = await detectedGradleError.handler(
+      line: detectedGradleErrorLine,
+      project: project,
+      usesAndroidX: usesAndroidX,
+      shouldBuildPluginAsAar: shouldBuildPluginAsAar,
+    );
+    if (status == GradleBuildStatus.RETRY && retries >= 1) {
+      await buildGradleApp(
+        project: project,
+        androidBuildInfo: androidBuildInfo,
+        target: target,
+        isBuildingBundle: isBuildingBundle,
+        gradleErrors: gradleErrors,
+        shouldBuildPluginAsAar: shouldBuildPluginAsAar,
+        retries: retries - 1,
+      );
+      return;
+    }
+    if (status == GradleBuildStatus.RETRY_WITH_AAR_PLUGINS &&
+        retries >= 1) {
+      await buildGradleApp(
+        project: project,
+        androidBuildInfo: androidBuildInfo,
+        target: target,
+        isBuildingBundle: isBuildingBundle,
+        gradleErrors: gradleErrors,
+        shouldBuildPluginAsAar: true,
+        retries: retries - 1,
+      );
+      return;
     }
+    throwToolExit(
+      'Gradle task $assembleTask failed with exit code $exitCode',
+      exitCode: exitCode,
+    );
   }
 
-  final FlutterManifest manifest = project.manifest;
+  flutterUsage.sendTiming('build', 'gradle', Duration(milliseconds: sw.elapsedMilliseconds));
```

Comment:
`sw.elapsed` is already a `Duration`.

You can also include a named string parameter 'label' if you'd like to somehow indicate that the timing includes a retry.